book/src/introduction.md

```diff
@@ -0,0 +1,168 @@
+# Introduction
+
+The JNI(Java Native Interface) allows Java code that runs inside a Java Virtual
+Machine (VM) to interoperate with applications and libraries written in other
+programming languages, such as C, C++, assembly, and Rust. This book is about
+writing Rust that works with Java over JNI. It's intended for people that know a
+bit of Rust and Java, and want to learn how to use them together in-process,
+hopefully without hitting their head against the wall. Suggestions, corrections
+and improvements are welcome and wanted.
+
+JNI comes with a lot of sharp edges. The native portion needs to be built for
+each platform you support, and JNI misuse can lead to resource leaks or JVM
+crashes. There are also performance implications of using JNI in certain ways.
+For example, Java Strings and Buffers sometimes have to be copied when they're
+used on the native side. Due to the drawbacks of JNI, it's often worth
+considering alternatives (like reimplementation in one language or IPC/RPC).
+Eventually, [Project Panama](https://openjdk.java.net/projects/panama/) may also
+offer a simpler or more efficient API.
+
+There are plenty of reasons why you should consider Rust if you need to write
+native code that works with Java. Most notably, you can separate JNI code into a
+specific layer, so that all benefits you get from Rust in normal application or
```

Comment:
We may also say that unlike C, jni-rs offers some safer to use abstractions.

For example, there are Executor that ensures native threads get detached when they are destroyed; GlobalRef/AutoLocal/AutoByteArray that will always delete the ref; with_local_frame that won't let you miss "Pop", etc. Most of them could be implemented equally well in C++, some safety features, however, are unique to Rust (like not being able to share JNIEnv/JObject across threads).

For C++, there is https://github.com/mapbox/jni.hpp, though we didn't use that, because we had to integrate a _Rust_ framework with Java :)

Reply:
Thanks, I've updated the tone and discussion here, so that it talks more about jni-rs abstractions. The later exercises/solutions should also include some discussion of the drop guarantees, since they are really useful.